dbms/src/TestUtils/executorSerializer.cpp

```diff
@@ -0,0 +1,276 @@
+            },
+            ", ");
+        buf.append(")");
+    }
+    else
+    {
+        buf.fmtAppend("<{}, {}>", decodeDAGInt64(expr.val()), getFieldTypeName(expr.field_type().tp()));
+    }
+}
+
+void serializeSelection(const String & executor_id, const tipb::Selection & sel, FmtBuffer & buf)
+{
+    buf.fmtAppend("{} | ", executor_id);
+    // currently only support "and" function in selection executor.
+    buf.joinStr(
+        sel.conditions().begin(),
+        sel.conditions().end(),
+        [&](const auto & expr, FmtBuffer &) {
+            serializeExpression(expr, buf);
+        },
+        " and ");
+    buf.append("}\n");
+}
+
+void serializeLimit(const String & executor_id, const tipb::Limit & limit, FmtBuffer & buf)
+{
+    buf.fmtAppend("{} | {}\n", executor_id, limit.limit());
+}
+
+void serializeProjection(const String & executor_id, const tipb::Projection & proj, FmtBuffer & buf)
+{
+    buf.fmtAppend("{} | {{", executor_id);
+    buf.joinStr(
+        proj.exprs().begin(),
+        proj.exprs().end(),
+        [&](const auto & expr, FmtBuffer &) {
+            serializeExpression(expr, buf);
+        },
+        ", ");
+    buf.append("}\n");
+}
+
+void serializeAggregation(const String & executor_id, const tipb::Aggregation & agg, FmtBuffer & buf)
+{
+    buf.fmtAppend("{} | group_by: {{", executor_id);
+    buf.joinStr(
+        agg.group_by().begin(),
+        agg.group_by().end(),
+        [&](const auto & group_by, FmtBuffer &) {
+            serializeExpression(group_by, buf);
+        },
+        ", ");
+    buf.append("}, agg_func: {");
+    buf.joinStr(
+        agg.agg_func().begin(),
+        agg.agg_func().end(),
+        [&](const auto & func, FmtBuffer &) {
+            serializeExpression(func, buf);
+        },
+        ", ");
+    buf.append("}\n");
+}
+
+void serializeTopN(const String & executor_id, const tipb::TopN & top_n, FmtBuffer & buf)
+{
+    buf.fmtAppend("{} | order_by: {{", executor_id);
+    buf.joinStr(
+        top_n.order_by().begin(),
+        top_n.order_by().end(),
+        [&](const auto & order_by, FmtBuffer & fb) {
+            fb.append("(");
+            serializeExpression(order_by.expr(), buf);
+            fb.fmtAppend(", desc: {})", order_by.desc());
+        },
+        ", ");
+    buf.fmtAppend("}}, limit: {}\n", top_n.limit());
+}
+
+void serializeJoin(const String & executor_id, const tipb::Join & join, FmtBuffer & buf)
+{
+    assert(join.left_join_keys_size() > 0);
+    assert(join.right_join_keys_size() > 0);
+    buf.fmtAppend("{} | {}, {}. left_join_keys: {{", executor_id, getJoinTypeName(join.join_type()), getJoinExecTypeName(join.join_exec_type()));
+    toString(join.left_join_keys(), buf);
+    buf.append("}, right_join_keys: {");
+    toString(join.right_join_keys(), buf);
+    buf.append("}\n");
+}
+
+void serializeExchangeSender(const String & executor_id, const tipb::ExchangeSender & sender, FmtBuffer & buf)
+{
+    assert(sender.all_field_types_size() > 0);
```

Comment:
nit: after wraping a function `toString`, `assert(sender.all_field_types_size() > 0);` is useless